### ISSUE TITLE
Use plugin manager inside books.stream to get missing book data 

### ISSUE THREAD

**Issue by fadwamahmoud** [librocco/librocco#322]
The plan is to adopt a plugin-based system for integrating different sources of book data (ie a plugin for each data source #311 ). The point is to expose a plugin manager to the app code which internally allows individual plugins (determined by the user) to handle fetching and data mapping of book data from different sources based on an ISBN and making this system extensible so that future data sources could be added as well. 

Where the plugin manager exposes this main function:

    getBook(isbn: string): Promise<BookMetadata>

So the flow that makes the most sense is that when the books are being scanned the isbns get fetched in the background and as the book data becomes available they start showing up in the note in their respective rows.

### Behaviour:

1. User scans  bunch of books.
2. The transactions show up in the note without the book data at first.
3. Plugin is triggered and executes in the background .
4. Book data shows up in the transaction rows.

### Implementation:

inside note.stream, For each ISBN, if the book data is not found in the database, call plugin manager to fetch the data from the plugins

// books.ts

```
const stream = (ctx: debug.DebugCtx, isbns: string[]) => {
  return new Observable<BookEntry[]>((subscriber) => {
    //...
    
    concat(initialState, changeStream)
      .pipe(
        tap(debug.log(ctx, "books_stream:result:raw")),
        map(unwrapDocs),
        tap(debug.log(ctx, "books_stream:result:transformed")),
        
        // for each book not found in the db get from plugins
        mergeMap((books: (BookEntry | undefined)[]) =>
          Promise.all(
            books.map((book: BookEntry | undefined) => {
              if (book !== undefined) {
                return observableOf(book);
              } else {
                return observableFrom(pluginManager.getBook(isbn)).pipe(
                  catchError(error => {
                    debug.error(ctx, `Error fetching book with isbn ${isbn} from plugins`, error);
                    return observableOf(undefined);
                  })
                );
              }
            })
          )
        ),
      )
      .subscribe((doc) => subscriber.next(doc));
      
    //...
  });
}
```

**Comment by ikusteu** [comment]
On point, except for the tiny detail. The book data fetched isn't persisted. It's a tiny detail in implementation, but a big improvement in the general data flow:

Let's leave the entries stream as is (completely). This is the final step of the current implementation:
```ts
	const tableData = entriesStream.pipe(
		switchMap(({ rows }) => {
			// Map rows to just isbns
			const isbns = rows.map((entry) => entry.isbn);

			debug.log(ctx, "display_entries_store:table_data:retrieving_books")({ isbns });

			// Return array of merged values of books and volume stock client
			return db
				.books()
				.stream(ctx, isbns)
				.pipe(
					tap(debug.log(ctx, "display_entries_store:table_data:retrieved_books")),
					map((booksFromDb) => booksFromDb.map((b = {} as BookEntry, i) => ({ ...b, ...rows[i] }))),
					tap(debug.log(ctx, "display_entries_store:table_data:merged_books"))
				);
		})
	);
```

In here, the stream that streams entries from the db (inventory only) is switched to a new stream, streaming book data from the db (with respect to shown entries) and merging the two (resulting in inventory data - isbn, quantity and book data). So, whenever the entries stream updates it will generate a new stream (merging entries and book data), but also whenever book data updates, it will stream those updates as the final result.

I suggest we contain the book data fetching to the `books` section of the db:
- `db.books().stream()` is called with the list of isbns (e.g. `1234567890`, `1234567891`, `1234567892`)
- `books` section finds data for `1234567890` and `1234567892`
- `books` section streams [book data for `1234567890`, `null`, book data for `1234567892`] to the UI immediately and fetches book data, using the book fetching plugin, for `1234567891`, in the background
- on response (fetch request) `books` section store the book data and streams updated entries: [book data for `1234567890`, book data for `1234567891`, book data for `1234567892`]

**Comment by ikusteu** [comment]
This is kinda included in #344 or refactored as different tickets
